base58/src/error.rs

```diff
@@ -123,3 +116,56 @@ impl fmt::Display for BadByteError {
 
 #[cfg(feature = "std")]
 impl std::error::Error for BadByteError {}
+
+/// Checksum was not correct.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct BadChecksumError {
+    /// The bad checksum.
+    bad: u32,
+    /// The expected checksum.
+    expected: u32,
+}
+
+impl BadChecksumError {
+    /// Creates a new [`BadChecksumError`] from `bad` byte.
+    pub fn new(bad: u32, expected: u32) -> Self { Self { bad, expected } }
+
+    /// Returns the bad checksum.
+    pub fn bad_checksum(&self) -> u32 { self.bad }
+
+    /// Returns the expected checksum.
+    pub fn expected_checksum(&self) -> u32 { self.expected }
+}
+
+impl fmt::Display for BadChecksumError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "base58 checksum {:#x} does not match expected {:#x}", self.bad, self.expected)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for BadChecksumError {}
+
+/// The decode base58 data was too short.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct TooShortError {
+    /// The length of the decoded data.
+    len: usize,
+}
+
+impl TooShortError {
+    /// Creates a new [`TooShortError`] from `len`.
+    pub fn new(len: usize) -> Self { Self { len } }
```

Comment:
This shouldn't be public.

Reply:
I'm undecided on this, I need to have another go at hiding `hex` (or I could hide `base58`) in `rust-bitcoin`, last time I tried with `hex` I wanted to create some of these types of error types (the struct hider ones) - that is how I stumbled upon the idea of getters/setters for the hider error types.

Reply:
If the constructor is public we're not really getting forward compatibility.

Reply:
We are getting forward compatibility as long as there are somewhat-sensible default values for any new fields. (And if there aren't we could make them `Option<T>`s and have the display use the old logic in the case that they're `None`)

Reply:
IMO it just complicates things to give users a feature almost nobody uses. (Though TBF I had an interesting case where it'd be helpful to be able to construct `core::str::Utf8Error` so these things do come up very rarely.)

Note that I'm not against publishing it *eventually* when we're confident we don't need to change it anymore, I'd just refrain from it in early releases (in the spirit of not increasing the API surface way too much).

Reply:
>almost nobody uses.

We literally can't get by without the ability to construct errors. If we have single-argument constructors then we can give these to `map_err` and our code is much smaller.

>Note that I'm not against publishing it eventually when we're confident we don't need to change it anymore, I'd just refrain from it in early releases (in the spirit of not increasing the API surface way too much).

Ok, this is fine. But in the meantime our own downstream error code is going to be harder to maintain. So I'm unsure that we're saving ourselves much maintenance burden. (Though perhaps we are minimizing the amount of API breakage.)

Reply:
> We literally can't get by without the ability to construct errors.

That doesn't require public constructors. Private are completely fine.

> But in the meantime our own downstream error code is going to be harder to maintain.

How? It just wraps the error.

> Though perhaps we are minimizing the amount of API breakage.

Yes, that's my point.